test/metabase/upload_test.clj

```diff
@@ -1412,6 +1416,59 @@
 
               (io/delete-file file))))))))
 
+
+(defn- mbql [mp table]
+  (let [table-metadata (lib.metadata/table mp (:id table))]
+    (lib/query mp table-metadata)))
+
+(defn- join-mbql [mp base-table join-table]
+  (let [base-table-metadata (lib.metadata/table mp (:id base-table))
+        join-table-metadata (lib.metadata/table mp (:id join-table))
+        ;; We use the primary keys as the join fields as we know they will exist and have compatible types.
+        pk-metadata         (fn [table]
+                              (let [field-id (t2/select-one-pk :model/Field
+                                                               :table_id (:id table)
+                                                               :semantic_type :type/PK)]
+                                (lib.metadata/field mp field-id)))
+        base-id-metadata         (pk-metadata base-table)
+        join-id-metadata         (pk-metadata join-table)]
+
+    (-> (lib/query mp base-table-metadata)
+        (lib/join (lib/join-clause join-table-metadata
+                                   [(lib/= (lib/ref base-id-metadata)
+                                           (lib/ref join-id-metadata))])))))
+
+(deftest ^:mb/once update-invalidate-model-cache-test
+  (mt/test-drivers (mt/normal-drivers-with-feature :uploads)
+    (doseq [action (actions-to-test driver/*driver*)]
+      (testing (action-testing-str action)
+        (with-upload-table! [table (create-upload-table!)]
+          (let [table-id    (:id table)
+                csv-rows    ["name" "Luke Skywalker"]
+                file        (csv-file-with csv-rows)
+                other-id    (mt/id :venues)
+                other-table (t2/select-one :model/Table other-id)
+                mp          (lib.metadata.jvm/application-database-metadata-provider (:db_id table))]
+
+            (mt/with-temp [:model/Card {question-id         :id} {:table_id table-id, :dataset_query (mbql mp table)}
+                           :model/Card {model-id            :id} {:table_id table-id, :type :model, :dataset_query (mbql mp table)}
+                           :model/Card {complex-model-id    :id} {:table_id table-id, :type :model :dataset_query (join-mbql mp table other-table)}
```

Comment:
super nit: these commas are inconsistent

Reply:
meta nit: Are super nits the opposite of pico nits? 😄